src/foam/core/CSV.js

```diff
@@ -0,0 +1,359 @@
+      class: 'String',
+      name: 'delimiter',
+      value: ','
+    },
+    {
+      class: 'String',
+      name: 'nestedObjectSeperator',
+      value: '__'
+    },
+    {
+      class: 'StringArray',
+      name: 'nestedObjectNames'
+    },
+    {
+      class: 'String',
+      name: 'nlStr',
+      value: '\n'
+    },
+    {
+      class: 'String',
+      name: 'undefinedStr',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'outputHeaderRow',
+      value: true
+    },
+    {
+      class: 'Function',
+      name: 'propertyPredicate',
+      value: function(o, p) { return ! p.transient; }
+    },
+    {
+      class: 'Boolean',
+      name: 'sortObjectKeys',
+      value: false
+    }
+  ],
+
+  methods: [
+    function out() {
+      for ( var i = 0 ; i < arguments.length ; i++ ) { 
+        this.buf_ += arguments[i];
+      }
+
+      return this;
+    },
+
+    function outputHeader(o) {
+      if ( ! this.outputHeaderRow ) return;
+
+      this.outputPropertyFilteredName(o);
+      this.removeTrailingComma().out(this.nlStr);
+    },
+
+    function removeTrailingComma() {
+      // Removes trailing comma
+      if ( this.buf_.charAt(this.buf_.length - 1) == ',' ) this.buf_ = this.buf_.slice(0, -1);
+      return this;
+    },
+
+    function stringify(o) {
+      // Outputs object headers
+      this.outputHeader(o);
+      var ret = this.buf_;
+      this.reset(); // reset to avoid retaining garbage
+
+      // Outputs object values
+      this.output(o);
+      this.removeTrailingComma().out(this.nlStr);
+      ret += this.buf_;
+      this.reset();
+
+      return ret;
+    },
+
+    function externalProperty(o, p) {
+      if ( ! this.propertyPredicate(o, p) ) return false;
+      if ( p.isDefaultValue(o[p.name]) ) return false;
+
+      return true;
+    },
+
+    function outputHeaderTitle(o) {
+      this.out(this.includeQuotes ? '"' : '').writePrefix().out(o, this.includeQuotes ? '"' : '', this.delimiter);
+    },
+
+    {
+      name: 'outputPropertyFilteredName',
+      code: foam.mmethod({
+        // Is there a `switch` like fall through so the function call isn't unecessarily repeated?
+        Undefined: function(o) { this.outputHeaderTitle(o); },
+        Null:      function(o) { this.outputHeaderTitle(o); },
+        String:    function(o) { this.outputHeaderTitle(o); },
+        Number:    function(o) { this.outputHeaderTitle(o); },
+        Boolean:   function(o) { this.outputHeaderTitle(o); },
+        Date:      function(o) { this.outputHeaderTitle(o); },
+        FObject:   function(o) {
+          // Gets and recurses through object properties
+          var ps = o.cls_.getAxiomsByClass(foam.core.Property);
+
+          for ( var i = 0 ; i < ps.length ; i++ ) {
+            this.outputPropertyName(o, ps[i]);
+          }
+        },
+        Array: function(o) { /* Ignore arrays in CSV */ },
+        Function: function(o) { /* Ignore functions in CSV */ },
+        Object: function(o) {
```

Comment:
I don't think CSV can support generic Objects since you don't know how many keys they're going to have so then you don't know which column means what.